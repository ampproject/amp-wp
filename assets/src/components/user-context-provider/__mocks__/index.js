/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { createContext } from '@wordpress/element';

export const User = createContext();

/**
 * MOCK.
 *
 * @param {Object}  props
 * @param {any}     props.children
 * @param {boolean} props.fetchingUser
 */
export function UserContextProvider({ children, fetchingUser }) {
	return (
		<User.Provider
			value={{
				savingDeveloperToolsOption: false,
				fetchingUser,
			}}
		>
			{children}
		</User.Provider>
	);
}
UserContextProvider.propTypes = {
	children: PropTypes.any,
	fetchingUser: PropTypes.bool,
};
