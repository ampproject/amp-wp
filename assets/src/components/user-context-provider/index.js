/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import {
	createContext,
	useEffect,
	useState,
	useRef,
	useCallback,
	useMemo,
	useContext,
} from '@wordpress/element';
import apiFetch from '@wordpress/api-fetch';

/**
 * Internal dependencies
 */
import { useAsyncError } from '../../utils/use-async-error';
import { Options } from '../options-context-provider';

export const User = createContext();

/**
 * Context provider user data.
 *
 * @param {Object}  props                                              Component props.
 * @param {?any}    props.children                                     Component children.
 * @param {boolean} props.onlyFetchIfPluginIsConfigured                Flag indicating whether the users data should be fetched only if the plugin is fully configured (i.e. the Onboarding Wizard has been completed).
 * @param {string}  props.userFieldReviewPanelDismissedForTemplateMode The key of the option to use from the settings endpoint.
 * @param {string}  props.userOptionDeveloperTools                     The key of the option to use from the settings endpoint.
 * @param {string}  props.usersResourceRestPath                        The REST path for interacting with the `users` resources.
 */
export function UserContextProvider({
	children,
	onlyFetchIfPluginIsConfigured = true,
	userFieldReviewPanelDismissedForTemplateMode,
	userOptionDeveloperTools,
	usersResourceRestPath,
}) {
	const { originalOptions, fetchingOptions } = useContext(Options);
	const { plugin_configured: pluginConfigured } = originalOptions;
	const [fetchingUser, setFetchingUser] = useState(false);
	const [developerToolsOption, setDeveloperToolsOption] = useState(null);
	const [
		reviewPanelDismissedForTemplateMode,
		setReviewPanelDismissedForTemplateMode,
	] = useState(null);
	const [originalDeveloperToolsOption, setOriginalDeveloperToolsOption] =
		useState(null);
	const [savingDeveloperToolsOption, setSavingDeveloperToolsOption] =
		useState(false);
	const [
		savingReviewPanelDismissedForTemplateMode,
		setSavingReviewPanelDismissedForTemplateMode,
	] = useState(false);
	const [didSaveDeveloperToolsOption, setDidSaveDeveloperToolsOption] =
		useState(false);
	const { setAsyncError } = useAsyncError();

	// This component sets state inside async functions. Use this ref to prevent state updates after unmount.
	const hasUnmounted = useRef(false);
	useEffect(
		() => () => {
			hasUnmounted.current = true;
		},
		[]
	);

	const hasDeveloperToolsOptionChange = useMemo(
		() =>
			null !== developerToolsOption &&
			developerToolsOption !== originalDeveloperToolsOption,
		[developerToolsOption, originalDeveloperToolsOption]
	);

	/**
	 * Fetch user options on mount.
	 */
	useEffect(() => {
		if (fetchingOptions) {
			return;
		}

		if (!pluginConfigured && onlyFetchIfPluginIsConfigured) {
			setOriginalDeveloperToolsOption(null);
			setDeveloperToolsOption(null);
			return;
		}

		if (
			!usersResourceRestPath ||
			fetchingUser ||
			null !== originalDeveloperToolsOption
		) {
			return;
		}

		/**
		 * Fetches user data from the REST endpoint for the current user.
		 */
		(async () => {
			setFetchingUser(true);

			try {
				const fetchedUser = await apiFetch({
					path: `${usersResourceRestPath}/me`,
				});

				if (true === hasUnmounted.current) {
					return;
				}

				setOriginalDeveloperToolsOption(
					fetchedUser[userOptionDeveloperTools]
				);
				setDeveloperToolsOption(fetchedUser[userOptionDeveloperTools]);

				if (userFieldReviewPanelDismissedForTemplateMode) {
					setReviewPanelDismissedForTemplateMode(
						fetchedUser[
							userFieldReviewPanelDismissedForTemplateMode
						]
					);
				}
			} catch (e) {
				setAsyncError(e);
				return;
			}

			setFetchingUser(false);
		})();
	}, [
		onlyFetchIfPluginIsConfigured,
		fetchingOptions,
		fetchingUser,
		originalDeveloperToolsOption,
		pluginConfigured,
		setAsyncError,
		userFieldReviewPanelDismissedForTemplateMode,
		userOptionDeveloperTools,
		usersResourceRestPath,
	]);

	/**
	 * Sends the dev tools option back to the REST endpoint to be saved.
	 */
	const saveDeveloperToolsOption = useCallback(async () => {
		if (!hasDeveloperToolsOptionChange) {
			return;
		}

		setSavingDeveloperToolsOption(true);

		try {
			const fetchedUser = await apiFetch({
				method: 'post',
				path: `${usersResourceRestPath}/me`,
				data: {
					[userOptionDeveloperTools]: developerToolsOption,
				},
			});

			if (true === hasUnmounted.current) {
				return;
			}

			setOriginalDeveloperToolsOption(
				fetchedUser[userOptionDeveloperTools]
			);
			setDeveloperToolsOption(fetchedUser[userOptionDeveloperTools]);
		} catch (e) {
			setAsyncError(e);
			return;
		}

		setDidSaveDeveloperToolsOption(true);
		setSavingDeveloperToolsOption(false);
	}, [
		hasDeveloperToolsOptionChange,
		developerToolsOption,
		setAsyncError,
		userOptionDeveloperTools,
		usersResourceRestPath,
	]);

	/**
	 * Sends the template mode for which the "Review" panel is dismissed back to
	 * the REST endpoint to be saved.
	 */
	const saveReviewPanelDismissedForTemplateMode = useCallback(
		async (templateMode) => {
			if (
				savingReviewPanelDismissedForTemplateMode ||
				!userFieldReviewPanelDismissedForTemplateMode
			) {
				return;
			}

			// Update the local state immediately.
			setReviewPanelDismissedForTemplateMode(templateMode);
			setSavingReviewPanelDismissedForTemplateMode(true);

			try {
				await apiFetch({
					method: 'post',
					path: `${usersResourceRestPath}/me`,
					data: {
						[userFieldReviewPanelDismissedForTemplateMode]:
							templateMode,
					},
				});

				if (true === hasUnmounted.current) {
					return;
				}
			} catch (e) {
				setAsyncError(e);
				return;
			}

			setSavingReviewPanelDismissedForTemplateMode(false);
		},
		[
			savingReviewPanelDismissedForTemplateMode,
			setAsyncError,
			userFieldReviewPanelDismissedForTemplateMode,
			usersResourceRestPath,
		]
	);

	return (
		<User.Provider
			value={{
				developerToolsOption,
				fetchingUser,
				didSaveDeveloperToolsOption,
				hasDeveloperToolsOptionChange,
				reviewPanelDismissedForTemplateMode,
				originalDeveloperToolsOption,
				saveDeveloperToolsOption,
				savingDeveloperToolsOption,
				setDeveloperToolsOption,
				saveReviewPanelDismissedForTemplateMode,
				savingReviewPanelDismissedForTemplateMode,
			}}
		>
			{children}
		</User.Provider>
	);
}

UserContextProvider.propTypes = {
	children: PropTypes.any,
	onlyFetchIfPluginIsConfigured: PropTypes.bool,
	userFieldReviewPanelDismissedForTemplateMode: PropTypes.string,
	userOptionDeveloperTools: PropTypes.string.isRequired,
	usersResourceRestPath: PropTypes.string.isRequired,
};
