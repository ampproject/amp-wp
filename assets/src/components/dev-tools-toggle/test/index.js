/**
 * External dependencies
 */
import { act } from 'react-dom/test-utils';
import { create } from 'react-test-renderer';

/**
 * WordPress dependencies
 */
import { render } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { DevToolsToggle } from '../';
import { UserContextProvider } from '../../user-context-provider';

jest.mock('../../user-context-provider');

let container;

describe('DevToolsToggle', () => {
	beforeEach(() => {
		container = document.createElement('div');
		document.body.appendChild(container);
	});

	afterEach(() => {
		document.body.removeChild(container);
		container = null;
	});

	it('matches snapshot', () => {
		const wrapper = create(
			<UserContextProvider>
				<DevToolsToggle />
			</UserContextProvider>
		);
		expect(wrapper.toJSON()).toMatchSnapshot();
	});

	it('renders a loading spinner when a user data is fetched', () => {
		act(() => {
			render(
				<UserContextProvider fetchingUser={true}>
					<DevToolsToggle />
				</UserContextProvider>,
				container
			);
		});
		expect(
			container.querySelector('.amp-spinner-container')
		).not.toBeNull();
	});

	it('matches snapshot for the loading state', () => {
		const wrapper = create(
			<UserContextProvider fetchingUser={true}>
				<DevToolsToggle />
			</UserContextProvider>
		);
		expect(wrapper.toJSON()).toMatchSnapshot();
	});

	it('can be toggled', () => {
		act(() => {
			render(
				<UserContextProvider>
					<DevToolsToggle />
				</UserContextProvider>,
				container
			);
		});
		expect(container.querySelector('input:checked')).toBeNull();

		act(() => {
			container
				.querySelector('input')
				.dispatchEvent(new global.MouseEvent('click'));
		});

		expect(container.querySelector('input:checked')).not.toBeNull();

		act(() => {
			container
				.querySelector('input')
				.dispatchEvent(new global.MouseEvent('click'));
		});

		expect(container.querySelector('input:checked')).toBeNull();
	});
});
