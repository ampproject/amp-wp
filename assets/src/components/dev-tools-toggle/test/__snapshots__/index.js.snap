// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevToolsToggle matches snapshot 1`] = `
<div
  className="amp-setting-toggle"
>
  <div
    className="components-base-control components-toggle-control css-wdf2ti-Wrapper e1puf3u4"
  >
    <div
      className="components-base-control__field css-10urnh1-StyledField-deprecatedMarginField e1puf3u3"
    >
      <span
        className="components-form-toggle"
      >
        <input
          checked={false}
          className="components-form-toggle__input"
          id="inspector-toggle-control-0"
          onChange={[Function]}
          type="checkbox"
        />
        <span
          className="components-form-toggle__track"
        />
        <span
          className="components-form-toggle__thumb"
        />
      </span>
      <label
        className="components-toggle-control__label"
        htmlFor="inspector-toggle-control-0"
      >
        <div
          className="amp-setting-toggle__label-text"
        >
          <h3>
            Enable Developer Tools
          </h3>
        </div>
      </label>
    </div>
  </div>
</div>
`;

exports[`DevToolsToggle matches snapshot for the loading state 1`] = `
<div
  className="amp-spinner-container"
>
  <svg
    className="components-spinner css-1igdw1q-StyledSpinner e1bj2jdf2"
    focusable="false"
    role="presentation"
    viewBox="0 0 100 100"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      className="css-mx768t-SpinnerTrack-commonPathProps e1bj2jdf1"
      cx="50"
      cy="50"
      r="50"
      vectorEffect="non-scaling-stroke"
    />
    <path
      className="css-7c7ebc-SpinnerIndicator-commonPathProps e1bj2jdf0"
      d="m 50 0 a 50 50 0 0 1 50 50"
      vectorEffect="non-scaling-stroke"
    />
  </svg>
</div>
`;
